robot/impl/resource_manager.go

```diff
@@ -830,20 +838,40 @@ func (manager *resourceManager) updateResources(
 	defer manager.configLock.Unlock()
 	var allErrs error
 
+	// stores resource names so that if there is duplicate we error
+	seenResources := make(map[string]bool)
+	for _, r := range manager.resources.Names() {
+		seenResources[r.String()] = true
+	}
+
 	for _, s := range conf.Added.Services {
 		rName := s.ResourceName()
 		if manager.opts.untrustedEnv && rName.API == shell.API {
 			allErrs = multierr.Combine(allErrs, errShellServiceDisabled)
 			continue
 		}
+		if err := manager.validateUniqueResource(seenResources, rName.String()); err != nil {
+			allErrs = multierr.Combine(allErrs, err)
+			continue
+		}
+
 		allErrs = multierr.Combine(allErrs, manager.markResourceForUpdate(rName, s, s.Dependencies()))
 	}
 	for _, c := range conf.Added.Components {
 		rName := c.ResourceName()
+		if err := manager.validateUniqueResource(seenResources, rName.String()); err != nil {
+			allErrs = multierr.Combine(allErrs, err)
+			continue
+		}
```

Comment:
```suggestion
		if _, ok := manager.resources.Node(rName); ok {
			return errors.Errorf("resource %s already exists in the config", r.Name())
		}
```
(and delete seenResources)

(unless I am missing something)

Reply:
This works for everything except for modules + processes (they have a diff process for managing resources so don't have resource Names) and have unique IDs instead. So for the sake of just keeping it all in sync, this function may be better even though i agree using Node() is cleaner.

Reply:
But you don't add module names to seenResources..? It is just the names of the resources in the graph

Is that a bug or am I misunderstanding your comment?

Reply:
(Also you're right that Node() might not work because I'm not sure if it adds the resources instantly after marking for update. Would have to read the code)

Reply:
I do check for module name and process ID in the resource graph. And yes also wasn't sure about that but decided not to even try it based on the process and module different resource types

Reply:
Yes, you are right that the module doesn't get added into the resource map at the top level if it isn't part of the same resource graph. But `validateUniqueResource` adds modules that are in the current config to the resource map when it is called on the module name, so if you start and you have 2 modules with the same name, it will error out.

Reply:
Config v1: 

```
modules: [
{ 
"name": "roxys-cool-module"
...
}
]
```

The user updates and triggers a reconfigure. They decide to upload the following config:
```
modules :[
{
"name":"roxys-cool-module"
...
},
{
"name":"roxys-cool-module"
...
}
]
```

Will reconfigure with: 
```
conf.Added.Modules = [{name:"roxys-cool-module"...}] // note: will only have one module entry
```
Then the code will run
```


	for _, mod := range conf.Added.Modules {
		// this is done in config validation but partial start rules require us to check again
		if err := mod.Validate(""); err != nil {
			manager.logger.Errorw("module config validation error; skipping", "module", mod.Name, "error", err)
			continue
		}
		}

		if err := manager.validateUniqueResource(seenResources, mod.Name); err != nil {
			allErrs = multierr.Combine(allErrs, err)
			continue
		}
```

Note that the pre-existing version of "roxys-cool-module" will not be in `seenResources` yet because only things in `manager.resources` are added to `seenResources`.

This will pass reconfiguration and now we will have two modules with the name "roxys-cool-module".  Similarly, if you try to restart the robot now, it will fail because it will try to add both and now there will be a collision.

Reply:
ah i see, you are completely right here - i thought I had tested this and it worked but clearly an overstep by me. I will add modules/processes in the same way too. Any objections to that?